src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -0,0 +1,338 @@
+                    appArguments = $"{appArguments} -e {key} {args[key]}";
+                }
+            }
+
+            string command = $"shell am instrument {appArguments} -w {apkName}";
+            if (string.IsNullOrEmpty(instrumentationClassName))
+            {
+                _log.LogInformation($"Starting default instrumentation class on {apkName} (exit code -1 == success)");
+            }
+            else
+            {
+                _log.LogInformation($"Starting instrumentation class '{instrumentationClassName}' on {apkName}");
+                command = $"{command}/{instrumentationClassName}";
+            }
+            _log.LogDebug($"Raw command: '{command}'");
+            DateTime started = DateTime.Now;
+            var result = RunAdbCommand(command);
+            DateTime returned = DateTime.Now;
+
+            _log.LogInformation($"Running instrumentation class {displayName} took {(returned - started).TotalSeconds} seconds");
+            _log.LogDebug(FormatProcessOutputs(result));
+            return result;
+        }
+
+        #endregion
+
+        #region Process runner helpers
+
+        private string FormatProcessOutputs((string StandardOutput, string StandardError, int ExitCode) result)
+        {
+            StringBuilder output = new StringBuilder();
+            output.AppendLine($"Exit code: {result.ExitCode}");
+            output.AppendLine($"Standard Output:{Environment.NewLine}{result.StandardOutput}");
+            if (!string.IsNullOrEmpty(result.StandardError))
+            {
+                output.AppendLine($"Standard Error:{Environment.NewLine}{result.StandardOutput}");
+            }
+            return output.ToString();
+        }
+
+        public (string StandardOutput, string StandardError, int ExitCode) RunAdbCommand(string command)
+        {
+            return RunAdbCommand(command, TimeSpan.FromMinutes(5));
+        }
+
+        public (string StandardOutput, string StandardError, int ExitCode) RunAdbCommand(string command, TimeSpan timeOut)
+        {
+            if (!File.Exists(_absoluteAdbExePath))
+            {
+                throw new FileNotFoundException($"Provided path for adb.exe was not valid ('{_absoluteAdbExePath}')");
+            }
+
+            _log.LogDebug($"Executing: '{_absoluteAdbExePath} {command}'");
+
+            ProcessStartInfo processStartInfo = new ProcessStartInfo
+            {
+                CreateNoWindow = true,
+                UseShellExecute = false,
+                WorkingDirectory = Path.GetDirectoryName(_absoluteAdbExePath),
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                FileName = _absoluteAdbExePath,
+                Arguments = command,
+            };
+            Process p = new Process() { StartInfo = processStartInfo };
+            StringBuilder standardOut = new StringBuilder();
+            StringBuilder standardErr = new StringBuilder();

```

Comment:
How large can be the stdOut? I have reached the memory limit before using StringBuilder, maybe a Stream is a better approach in those cases where the stdout is large.

Reply:
this is straight up taken from mono runner code, let's cross that giant river when we get to it.